Foundation/NSURLSession/HTTPMessage.swift

```diff
@@ -0,0 +1,367 @@
+    ///
+    /// This can be converted into an `NSHTTPURLResponse`.
+    internal struct _ResponseHeaderLines {
+        let lines: [String]
+        init() {
+            self.lines = []
+        }
+        init(headerLines: [String]) {
+            self.lines = headerLines
+        }
+    }
+}
+
+extension URLSessionTask._ParsedResponseHeader {
+    /// Parse a header line passed by libcurl.
+    ///
+    /// These contain the <CRLF> ending and the final line contains nothing but
+    /// that ending.
+    /// - Returns: Returning nil indicates failure. Otherwise returns a new
+    ///     `ParsedResponseHeader` with the given line added.
+    @warn_unused_result
+    func byAppending(headerLine data: UnsafeBufferPointer<Int8>) -> URLSessionTask._ParsedResponseHeader? {
+        // The buffer must end in CRLF
+        guard
+            2 <= data.count &&
+                data[data.endIndex - 2] == CR &&
+                data[data.endIndex - 1] == LF
+            else { return nil }
+        let lineBuffer = UnsafeBufferPointer(start: UnsafePointer<UInt8>(data.baseAddress), count: data.count - 2)
+        guard let line = String(utf8Buffer: lineBuffer) else { return nil }
+        return byAppending(headerLine: line)
+    }
+    /// Append a status line.
+    ///
+    /// If the line is empty, it marks the end of the header, and the result
+    /// is a complete header. Otherwise it's a partial header.
+    /// - Note: Appending a line to a complete header results in a partial
+    ///     header with just that line.
+    @warn_unused_result
+    private func byAppending(headerLine line: String) -> URLSessionTask._ParsedResponseHeader {
+        if line.isEmpty {
+            switch self {
+            case .partial(let header): return .complete(header)
+            case .complete: return .partial(URLSessionTask._ResponseHeaderLines())
+            }
+        } else {
+            let header = partialResponseHeader
+            return .partial(header.byAppending(headerLine: line))
+        }
+    }
+    private var partialResponseHeader: URLSessionTask._ResponseHeaderLines {
+        switch self {
+        case .partial(let header): return header
+        case .complete: return URLSessionTask._ResponseHeaderLines()
+        }
+    }
+}
+private extension URLSessionTask._ResponseHeaderLines {
+    /// Returns a copy of the lines with the new line appended to it.
+    @warn_unused_result
+    func byAppending(headerLine line: String) -> URLSessionTask._ResponseHeaderLines {
+        var l = self.lines
+        l.append(line)
+        return URLSessionTask._ResponseHeaderLines(headerLines: l)
+    }
+}
+internal extension URLSessionTask._ResponseHeaderLines {
+    /// Create an `NSHTTPRULResponse` from the lines.
+    ///
+    /// This will parse the header lines.
+    /// - Returns: `nil` if an error occured while parsing the header lines.
+    @warn_unused_result
+    func createHTTPURLResponse(for URL: URL) -> NSHTTPURLResponse? {
```

Comment:
Not yet, though it's listed [here](https://github.com/apple/swift-evolution/blob/master/proposals/0086-drop-foundation-ns.md). If there was no reason for not dropping it, I can get it done.